### PR TITLE
update resume support to draft/resume-0.3

### PR DESCRIPTION
**Pull request by slingamn**
cc @jesopo

This implements https://github.com/DanielOaks/ircv3-specifications/blob/0844bc2cb7e906ca9dceecab5033bafdbbca48ba/extensions/resume.md

It went quite well!

By design, a successful resume bypasses all auth checks (`PASS`, sasl-required, etc.). Formerly, client lookup by nickname was enough to ensure that the old client had successfully registered (and therefore passed all the auth checks). This implementation needs to check explicitly; this may or may not be worth noting in the spec.

### CODE REVIEW

**Comment by slingamn** [comment]
Corresponding irctest change: https://github.com/DanielOaks/irctest/pull/17

File: irc/handlers.go [hunk 1]
```diff
@@ -2258,28 +2258,26 @@ func renameHandler(server *Server, client *Client, msg ircmsg.IrcMessage, rb *Re
...
 	var timestamp time.Time
-	if 2 < len(msg.Params) {
-		ts, err := time.Parse(IRCv3TimestampFormat, msg.Params[2])
+	if 1 < len(msg.Params) {
+		ts, err := time.Parse(IRCv3TimestampFormat, msg.Params[1])
 		if err == nil {
 			timestamp = ts
 		} else {
-			rb.Add(nil, server.name, ERR_CANNOT_RESUME, oldnick, client.t("Timestamp is not in 2006-01-02T15:04:05.999Z format, ignoring it"))
+			rb.Add(nil, server.name, "RESUME", "ERR", client.t("Timestamp is not in 2006-01-02T15:04:05.999Z format, ignoring it"))
```
Comment (DanielOaks) [inline irc/handlers.go]: let's make this an `ERR WARN`

File: irc/resume.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	rm.RLock()
+	defer rm.RUnlock()
+
+	id := token[:utils.SecretTokenLength]
+	pair, ok := rm.resumeIDtoCreds[id]
+	if ok {
+		if utils.SecretTokensMatch(pair.secret, token[utils.SecretTokenLength:]) {
+			// disallow resume of an unregistered client; this prevents the use of
+			// resume as an auth bypass
+			if pair.client.Registered() {
```
Comment (DanielOaks) [inline irc/resume.go]: really nice catch here

**Comment by DanielOaks** [comment]
With that downgrade of the timestamp message to a `WARN`, I'm 100% happy for this to be merged. Awesome work mate.

**Pull request merged**